DOCS-9094: Fixed when only a single flow instances is allowed

ethanport opened the pull request:
Rewrote and clarified. Added details about using maxconcurrency to decide when only one flow instance can execute. Clarified triggered vs. executed language.  The existing bullet about when only one flow instance is allowed is wrong.

ethanport committed:
Fixed when only a single flow instances is allowed

Rewrote and clarified. Added details about using maxconcurrency to decide when only one flow instance can execute. Clarified triggered vs. executed language.  The existing bullet about when only one flow instance is allowed is wrong.

dukesphere commented on modules/ROOT/pages/scheduler-concept.adoc:
```diff
@@ -3,16 +3,18 @@ ifndef::env-site,env-github[]
 endif::[]
 
-The Scheduler component enables you to trigger a flow when a time-based condition is met. You can configure it to be triggered at a regular interval or give it a more flexible cron expression. For example, a Scheduler might trigger an event to start a flow every 5 seconds.
+The Scheduler is a flow event source that is configured to trigger at certain times to execute the Scheduler's flow. A Scheduler is either configured with a fixed repeating triggering *frequency* (such as every second), or with a more general cron *expression*. A cron expression can specify one or more specific dates and times, but more often specifies recurring dates, times, or intervals (for ex ...
```
```suggestion
The Scheduler component is an event source in a flow that initiates execution of that flow at configured times. You can configure a fixed and repeating frequency (such as every second) or a cron expression. A cron expression can specify one or more specific dates and times, but more commonly, the expression specifies recurring dates, times, or intervals. For example, the cron expression `* * * ? * *` configures a Scheduler to start its flow every second. 
```

dukesphere commented on modules/ROOT/pages/scheduler-concept.adoc:
```diff
@@ -3,16 +3,18 @@ ifndef::env-site,env-github[]
-The Scheduler component enables you to trigger a flow when a time-based condition is met. You can configure it to be triggered at a regular interval or give it a more flexible cron expression. For example, a Scheduler might trigger an event to start a flow every 5 seconds.
+The Scheduler is a flow event source that is configured to trigger at certain times to execute the Scheduler's flow. A Scheduler is either configured with a fixed repeating triggering *frequency* (such as every second), or with a more general cron *expression*. A cron expression can specify one or more specific dates and times, but more often specifies recurring dates, times, or intervals (for ex ...
+
+The flow can be configured to only allow the Scheduler event source to execute one flow instance at a time, or can allow concurrent flow instances to execute. 
```
```suggestion
You can configure the Flow component to run concurrent instances of the flow or to prevent the Scheduler from starting more than one instance of the flow at a time. 
```

dukesphere commented on modules/ROOT/pages/scheduler-concept.adoc:
```diff
@@ -3,16 +3,18 @@ ifndef::env-site,env-github[]
 * If xref:execution-engine.adoc#backpressure[back-pressure] is triggered because no resources are available at the time of the trigger, that trigger is skipped.
 //Enhancement request for this: MULE-14930
-* In a Mule runtime engine cluster or multi-worker CloudHub deployment, the Scheduler runs (or triggers the flow) only on the primary node (that is, only in one Mule instance).
+* If the flow is configured with `maxconcurrency` set to 1, then while a triggered flow is still executing, subsequent Scheduler triggers cause the Scheduler's back-pressure to silently skip the trigger and no new flow instances are executed. 
```
```suggestion
* If the `maxconcurrency` setting of a Flow component is set to `1`, no new flow instances will execute while the flow is executing. The Scheduler's back-pressure silently skips triggers while the flow is executing. 
```

dukesphere replied:
This suggested change was replaced with another below.

dukesphere commented on modules/ROOT/pages/scheduler-concept.adoc:
```diff
@@ -30,18 +32,18 @@ To download and open an example project while you are in Anypoint Studio, click
 
 | *Start Delay*
-| The amount of time to wait before triggering the flow for the first time after the application is started. This value is expressed in the same unit of time as the frequency.
+| The amount of time to wait before triggering execution of its flow for the first time after the application is started. This value is expressed in the same unit of time as the frequency.
```
```suggestion
| The amount of time to wait before triggering the execution of its flow for the first time after the application starts. This value is expressed in the same unit of time as the frequency.
```

dukesphere commented on modules/ROOT/pages/scheduler-concept.adoc:
```diff
@@ -53,13 +55,13 @@ MILLISECONDS, SECONDS, MINUTES, HOURS, DAYS
 == Cron Expressions
 
-Cron is a widely used standard for describing time and date information.
+Cron is a widely used standard for describing time and date information and repeating intervals.
```
```suggestion
Cron is a widely used standard for describing date-time information and repeating intervals.
```

dukesphere commented on modules/ROOT/pages/scheduler-concept.adoc:
```diff
@@ -53,13 +55,13 @@ MILLISECONDS, SECONDS, MINUTES, HOURS, DAYS
 
 The Scheduler keeps track of every second and creates a Mule event when the
-Quartz Cron expression matches your time-date setting. You can trigger the event
+Quartz Cron expression matches your time-date setting. You can configure the Scheduler to trigger
```
```suggestion
Quartz cron expression matches your date-time setting. You can configure the Scheduler to trigger
```

dukesphere commented on modules/ROOT/pages/scheduler-concept.adoc:
```diff
@@ -53,13 +55,13 @@ MILLISECONDS, SECONDS, MINUTES, HOURS, DAYS
 just once or at regular intervals.
 
-A date-time expression consists of six required settings and can include the
+A *date-time expression* consists of six required settings and can include the
```
```suggestion
A date-time expression consists of six required settings and can include the
```

dukesphere commented on modules/ROOT/pages/scheduler-concept.adoc:
```diff
@@ -126,11 +128,13 @@ Java equivalents, such as `America/Los_Angeles` and
 
-In the Studio *Debug Mode* view, the Mule app disables the triggering of events. You can
-instead execute the Scheduler component when debugging by clicking the green arrow icon
+In the Studio *Debug Mode* view, the Mule app disables triggering of Schedulers. When debugging, you can
```
```suggestion
In the Studio *Debug Mode* view, a Mule application disables triggers on Scheduler components. When debugging, you can
```

dukesphere replied:
We need to verify whether it's true that Studio disables triggers when in debug mode.

ethanport replied:
It does not disable the Scheduler

dukesphere replied:
Yep, I'm finding that it only stops if you stop the flow by adding a break point or if you actively select Stop Scheduler from the right-click menu on the Scheduler...

dukesphere commented on modules/ROOT/pages/scheduler-concept.adoc:
```diff
@@ -126,11 +128,13 @@ Java equivalents, such as `America/Los_Angeles` and
-In the Studio *Debug Mode* view, the Mule app disables the triggering of events. You can
-instead execute the Scheduler component when debugging by clicking the green arrow icon
+In the Studio *Debug Mode* view, the Mule app disables triggering of Schedulers. When debugging, you can
+instead trigger a Scheduler component when debugging by clicking the green arrow icon
```
```suggestion
instead trigger a Scheduler component by clicking the green arrow icon
```

dukesphere reviewed (commented):
suggested changes

ethanport committed:
Added back pressure error type and scenarios

When back pressure is triggered by a Scheduler, an error is thrown and the trigger is ignored. We verified this with an actual test in Studio.

dukesphere commented on modules/ROOT/pages/scheduler-concept.adoc:
```diff
@@ -11,9 +11,9 @@ Consider the following when you implement a Scheduler in your Mule application:
 * Schedulers use the same timezone as the machine on which Mule is running. However, if an application is running in CloudHub, the Scheduler conforms to the UTC timezone, regardless of the geographic region in which the application is running.
 // TODO: MULE-14438
-* If xref:execution-engine.adoc#backpressure[back-pressure] is triggered because no resources are available at the time of the trigger, that trigger is skipped.
+* If xref:execution-engine.adoc#backpressure[back-pressure] is triggered because no resources are available at the time of the trigger, that trigger is skipped. In this case a *MULE:FLOW_BACK_PRESSURE* error is thrown indicating the flow cannot accept new events at this time.  
```
```suggestion
* If xref:execution-engine.adoc#backpressure[back-pressure] is triggered because no resources are available at the time of the trigger, the application skips the trigger and throws a MULE:FLOW_BACK_PRESSURE error. The error indicates that the flow cannot accept new events at that time.  
```

dukesphere replied:
moved change to new PR: https://github.com/mulesoft/docs-mule-runtime/pull/1650

dukesphere commented on modules/ROOT/pages/scheduler-concept.adoc:
```diff
@@ -11,9 +11,9 @@ Consider the following when you implement a Scheduler in your Mule application:
+* If xref:execution-engine.adoc#backpressure[back-pressure] is triggered because no resources are available at the time of the trigger, that trigger is skipped. In this case a *MULE:FLOW_BACK_PRESSURE* error is thrown indicating the flow cannot accept new events at this time.  
 //Enhancement request for this: MULE-14930
-* If the flow is configured with `maxconcurrency` set to 1, then while a triggered flow is still executing, subsequent Scheduler triggers cause the Scheduler's back-pressure to silently skip the trigger and no new flow instances are executed. 
+* If the flow is configured with `maxconcurrency` set to 1, then while a triggered flow is still executing, subsequent Scheduler triggers cause the Scheduler's back-pressure to skip the trigger and no new flow instances are executed. This guarantees that only one instance of the flow is executing in the Mule runtime at a particular time. 
```
```suggestion
* If the `maxconcurrency` attribute of a Flow component is set to `1`, then only one flow instance can execute at a time in the Mule runtime. In this case, The Mule runtime prevents new flow instances from executing while the first flow instance is still executing. Instead, the Scheduler's back-pressure does *not* queue up additional events. Instead, the flow skips new events created by the triggered Scheduler while the first flow is still executing. Once the current flow terminates, the next event triggered by the Scheduler will execute a new flow instance, and then the same blocking and event skipping pattern repeats. 
```

dukesphere reviewed (commented):
suggested style changes

ethanport committed:
Added how to enable debug logs

Debug level log messages will report every time the Scheduler triggers

ethanport committed:
Added how to enable debug logs

Debug level log messages will report every time the Scheduler triggers

dukesphere commented on modules/ROOT/pages/scheduler-concept.adoc:
```diff
@@ -3,16 +3,18 @@ ifndef::env-site,env-github[]
 endif::[]
 
-The Scheduler component enables you to trigger a flow when a time-based condition is met. You can configure it to be triggered at a regular interval or give it a more flexible cron expression. For example, a Scheduler might trigger an event to start a flow every 5 seconds.
+The Scheduler is a flow event source that is configured to trigger at certain times to execute the Scheduler's flow. A Scheduler is either configured with a fixed repeating triggering *frequency* (such as every second), or with a more general cron *expression*. A cron expression can specify one or more specific dates and times, but more often specifies recurring dates, times, or intervals (for ex ...
```
```suggestion
The Scheduler component is an event source in a flow that initiates execution of that flow at configured times. You can configure a fixed and repeating frequency (such as every second) or a cron expression. A cron expression can specify one or more specific dates and times, but more commonly, the expression specifies recurring dates, times, or intervals. For example, the cron expression `* * * ? * *` configures a Scheduler to start its flow every second. 
```

dukesphere replied:
moved change to new PR: https://github.com/mulesoft/docs-mule-runtime/pull/1650

dukesphere commented on modules/ROOT/pages/scheduler-concept.adoc:
```diff
@@ -3,16 +3,18 @@ ifndef::env-site,env-github[]
-The Scheduler component enables you to trigger a flow when a time-based condition is met. You can configure it to be triggered at a regular interval or give it a more flexible cron expression. For example, a Scheduler might trigger an event to start a flow every 5 seconds.
+The Scheduler is a flow event source that is configured to trigger at certain times to execute the Scheduler's flow. A Scheduler is either configured with a fixed repeating triggering *frequency* (such as every second), or with a more general cron *expression*. A cron expression can specify one or more specific dates and times, but more often specifies recurring dates, times, or intervals (for ex ...
+
+The flow can be configured to only allow the Scheduler event source to execute one flow instance at a time, or can allow concurrent flow instances to execute. 
```
```suggestion
A `maxConcurrency` setting on the Flow component affects the flow's Scheduler. The setting limits the number of instances of the flow that can run at the same time. Setting the value to `1` prevents the Scheduler from triggering more than one execution of the flow at a time. 
```

dukesphere replied:
moved change to new PR: https://github.com/mulesoft/docs-mule-runtime/pull/1650

ethanport committed:
Update modules/ROOT/pages/scheduler-concept.adoc

Co-authored-by: Duke <dukesphere@users.noreply.github.com>

ethanport committed:
Correct working for customer-hosted cluster

It's a customer-hosted Mule runtime cluster. We don't say Mule runtime engine.

ethanport committed:
Correct working for customer-hosted cluster

It's a customer-hosted Mule runtime cluster. We don't say Mule runtime engine.

ethanport reviewed (commented):
Also corrected maxConcurrency spelling, added it to the example, and removed doc:id attributes from the example (they just clutter the examples)

elrodro83 reviewed (changes requested):
There should be mention to the `disallow-concurrent-execution` flag here.
It is internally documented as: 

> If set to 'true', executions triggered while the flow is still running from a previous trigger will be ignored, effectively avoiding different triggers of the flow to run concurrently.

dukesphere commented on modules/ROOT/pages/scheduler-concept.adoc:
```diff
@@ -3,16 +3,18 @@ ifndef::env-site,env-github[]
 // TODO: MULE-14438
-* The scheduler component can trigger only one flow at any given time, and if a triggered flow is executing, it won't be triggered again until it finishes processing.
-* If xref:execution-engine.adoc#backpressure[back-pressure] is triggered because no resources are available at the time of the trigger, that trigger is skipped.
+* If xref:execution-engine.adoc#backpressure[back-pressure] is triggered because no resources are available at the time of the trigger, that trigger is skipped. In this case a *MULE:FLOW_BACK_PRESSURE* error is thrown indicating the flow cannot accept new events at this time. In Anypoint Studio, this error is only logged once the first time it occurs. You can capture log messages each time the Sc ...
```
```suggestion
* If xref:execution-engine.adoc#backpressure[back pressure] occurs because Mule lacks sufficient resources to process an event in the flow, the application skips the scheduled trigger for the event and throws a MULE:FLOW_BACK_PRESSURE error. This error indicates that the flow cannot accept new events at that time. Anypoint Studio logs this error only once, the first time it occurs. 
+
To capture log messages each time the Scheduler attempts to trigger an event, you can enable the xref:logging-in-mule#change-the-log-level[DEBUG] logging level in the Mule application. 
```

dukesphere replied:
moved change to new PR: https://github.com/mulesoft/docs-mule-runtime/pull/1650

dukesphere commented on modules/ROOT/pages/scheduler-concept.adoc:
```diff
@@ -3,16 +3,18 @@ ifndef::env-site,env-github[]
+* If xref:execution-engine.adoc#backpressure[back-pressure] is triggered because no resources are available at the time of the trigger, that trigger is skipped. In this case a *MULE:FLOW_BACK_PRESSURE* error is thrown indicating the flow cannot accept new events at this time. In Anypoint Studio, this error is only logged once the first time it occurs. You can capture log messages each time the Sc ...
 //Enhancement request for this: MULE-14930
-* In a Mule runtime engine cluster or multi-worker CloudHub deployment, the Scheduler runs (or triggers the flow) only on the primary node (that is, only in one Mule instance).
+* If the `maxConcurrency` attribute of a Flow component is set to `1`, then only one flow instance can execute at a time in the Mule runtime. In this case, The Mule runtime prevents new flow instances from executing while the first flow instance is still executing. Instead, the Scheduler's back-pressure does *not* queue up additional events. Instead, the flow skips new events created by the trigg ...
```
```suggestion
* If you set the `maxConcurrency` attribute of a Flow component to `1`, only one instance of the flow can execute at a time. When an instance is running, the application skips and avoids queuing any new events that are triggered by the Scheduler. Once that instance of the flow terminates and the Scheduler triggers an event, a new instance of the flow executes, and the same blocking and event-skipping pattern occurs. 
+
See also, `disallowConcurrentExecution` in xref:scheduler-xml-reference.adoc#scheduler-properties[<scheduler> Properties]. 
```

dukesphere replied:
moved change to new PR: https://github.com/mulesoft/docs-mule-runtime/pull/1650

dukesphere commented on modules/ROOT/pages/scheduler-concept.adoc:
```diff
@@ -3,16 +3,18 @@ ifndef::env-site,env-github[]
 //Enhancement request for this: MULE-14930
-* In a Mule runtime engine cluster or multi-worker CloudHub deployment, the Scheduler runs (or triggers the flow) only on the primary node (that is, only in one Mule instance).
+* If the `maxConcurrency` attribute of a Flow component is set to `1`, then only one flow instance can execute at a time in the Mule runtime. In this case, The Mule runtime prevents new flow instances from executing while the first flow instance is still executing. Instead, the Scheduler's back-pressure does *not* queue up additional events. Instead, the flow skips new events created by the trigg ...
+* In a multi-worker CloudHub deployment, or in a custoemr-hosted Mule runtime cluster, the Scheduler runs (and then executes the flow) only on the primary node (that is, only in one particular Mule runtime).
```
```suggestion
* In a multi-worker CloudHub deployment or in a customer-hosted cluster of Mule runtime engines, the Scheduler runs and executes the flow on the primary node only. That is, it runs on one specific Mule runtime engine and no others.
```

dukesphere replied:
moved change to new PR: https://github.com/mulesoft/docs-mule-runtime/pull/1650

dukesphere commented on modules/ROOT/pages/scheduler-concept.adoc:
```diff
@@ -53,13 +55,13 @@ MILLISECONDS, SECONDS, MINUTES, HOURS, DAYS
 just once or at regular intervals.
 
-A date-time expression consists of six required settings and can include the
+A *date-time expression* consists of six required settings and can include the
```
```suggestion
A date-time expression consists of six required settings and can include the
```

dukesphere replied:
note that things in bold do not get translated

dukesphere replied:
moved change to new PR: https://github.com/mulesoft/docs-mule-runtime/pull/1650

dukesphere commented on modules/ROOT/pages/scheduler-concept.adoc:
```diff
@@ -98,17 +100,17 @@ Note that the Scheduler component also supports Quartz Scheduler special charact
 //source info: +http://www.quartz-scheduler.org/documentation/quartz-2.x/tutorials/crontrigger.html+
 
-This example logs the message "hello" every second:
+This example logs the message "hello" every second, and only allows one flow instance to execute at a time:
```
```suggestion
This example logs the message "hello" every second and allows only one instance of the flow to execute at a time:
```

dukesphere replied:
moved change to new PR: https://github.com/mulesoft/docs-mule-runtime/pull/1650

dukesphere commented on modules/ROOT/pages/scheduler-concept.adoc:
```diff
@@ -124,13 +126,12 @@ Java equivalents, such as `America/Los_Angeles` and
-In the Studio *Debug Mode* view, the Mule app disables the triggering of events. You can
-instead execute the Scheduler component when debugging by clicking the green arrow icon
-that appears at the bottom of the component.
+When you debug a Mule application in Anypoint Studio, each Scheduler event source appears with an arrow icon underneath to remind you that the Scheduler can be stopped and started while the Mule application is being debugged. To stop or start the Scheduler event source, right click on the Scheduler and select Stop Scheduler or Start Scheduler from the Scheduler's right-click menu.  
```
```suggestion
When you run a Mule application in Anypoint Studio using debug mode, an arrow icon appears at the bottom of the Scheduler component to indicate that you can start or stop it from the Scheduler's right-click menu.

By default, the Scheduler runs when in debug mode. If you place a break point on a message processor in the flow, the Scheduler component will start the flow but will not run again until you remove or skip all break points from the flow.
```

dukesphere replied:
moved change to new PR: https://github.com/mulesoft/docs-mule-runtime/pull/1650

dukesphere commented on modules/ROOT/pages/scheduler-concept.adoc:
```diff
@@ -124,13 +126,12 @@ Java equivalents, such as `America/Los_Angeles` and
 //source info: +https://docs.oracle.com/javase/8/docs/api/java/time/ZoneId.html+
 
-== Debugging
+== Debugging a Scheduler
```
```suggestion
== Stop or Start a Scheduler in Debug Mode
```

dukesphere replied:
moved change to new PR: https://github.com/mulesoft/docs-mule-runtime/pull/1650

dukesphere reviewed (commented):
@ethanport please commit the suggestions you are good with. Let's chat about any others. Thx!

dukesphere committed:
DOCS-9094-schedulerPR-duke

dukesphere committed:
DOCS-9094-schedulerPR-duke performance

dukesphere committed:
DOCS-9094-schedulerPR-duke first sentence

dukesphere committed:
DOCS-9094-schedulerPR-duke consolidate maxConcurrency

dukesphere committed:
DOCS-9094 from convos w/ Martin and Rodro

dukesphere committed:
DOCS-9094 more from MB

dukesphere committed:
Merge pull request #1650 from mulesoft/DOCS-9094-schedulerPR-duke

Docs 9094 scheduler pr duke

dukesphere commented:
This is all addressed, including disallow-concurrent-execution through https://github.com/mulesoft/docs-mule-runtime/pull/2389
